WYSIWYG- und File-Upload-Element

MHajoha opened the pull request:
## Features

### `file_upload_element`

Wird mithilfe von [`MoodleQuickForm_filemanager`](https://moodledev.io/docs/5.0/apis/subsystems/form/usage/files#file-manager) realisiert. Für Uploads einzelner Dateien gäbe es auch die Möglichkeit, den simpleren `filepicker` zu verwenden, ich verwende der Einfachheit halber aber nur `filemanager`.

Der Server / das Paket erhält Form-Daten wie folgt:

```json
"mein_upload_feld": [
  {
    "filename": "Bitrot_2.png",
    "file_ref": "abb0a28d748f4cf4676b92a7e40047228bb98901-0457ca0fb8d22be8",
    "uploaded_at": "2025-09-29T16:19:19+00:00",
    "mime_type": "image/png"
  }
]
```

Der Dateiname existiert dabei ausschließlich in den Form-Daten (und dann vermutlich im Question State), Moodle-seitig wird der Dateiname "vergessen". Das gibt dem Paket die Freiheit, Dateien umzubenennen, ohne QPy-URLs zu brechen.

Die File-Ref besteht aus dem SHA-1 content hash, den Moodle schon berechnet, und einem XXH3 (schnell, gut verteilt, nicht-kryptografisch) Metadata-Hash über context id, user id, Autor, Lizenz und Quelle (Letztere drei sind Freitextfelder im file manager). Der Metadata-hash stellt sicher, die 5 Informationen erhalten bleiben, wenn eine inhaltsmäßig identische Datei mehrmals in einer Frage hochgeladen wird.

Grundsätzlich könnte man auch beides zusammenhashen, ich war mir aber unsicher, wie sehr das das Kollisionsrisiko erhöht, und dass man erkennen kann, welche Dateien identisch sind, ist ein netter Nebeneffekt.

Die meiste Komplexität stammt aus dem hin- und herschieben zwischen Draft-Area und permanenter Filearea. Während des Bearbeitens (also auch während des Validierens) liegen die Dateien in einer Draft-Area, wobei jedes Element seine eigene Draft-Area hat. Beim Speichern werden alle Dateien aus allen verwendeten Draft-Areas gemeinsam unter ihrer File-Ref in der permanenten File-Area gespeichert, von der jede Frage genau eine hat. 

Beim Laden einer Frage müssen umgekehrt wieder genau die Dateien in Draft-Areas kopiert werden, die Laut dem Paket zu den einzelnen Elementen gehören.

Das macht es u.A. möglich, eine Datei zwischen Elementen zu bewegen, zu kopieren, und mehrmals zu verwenden.

### `wysiwyg_editor_element`

Wird mithilfe von [`MoodleQuickForm_editor`](https://moodledev.io/docs/5.0/apis/subsystems/form/usage/files#editors) realisiert.

Der Server / das Paket erhält Form-Daten wie folgt:

```json
"mein_editor_feld": {
  "markup": "## Blablablabla\n\n![Bitrot](qpy://options/66fbfcdfbdb161b86f63e92723ab63f080d62fcb-9324cd2451433273)",
  "markup_format": "markdown",
  "html": "<h2>Blablablabla<\/h2>\r\n<p><\/p>\r\n<p><img class=\"img-fluid\" src=\"qpy:\/\/options\/66fbfcdfbdb161b86f63e92723ab63f080d62fcb-9324cd2451433273\" alt=\"Bitrot\" width=\"1536\" height=\"1024\"><\/p>",
  "files": [
    {
      "filename": "Bitrot_1.png",
      "file_ref": "66fbfcdfbdb161b86f63e92723ab63f080d62fcb-9324cd2451433273",
      "uploaded_at": "2025-09-29T16:22:25+00:00",
      "mime_type": "image/png"
    }
  ]
}
```

Dabei wird die HTML-"Übersetzung" explizit durch's Paket angefordert. Dateien funktionieren genauso wie im `file_upload_element`.

Liefert der Editor direkt HTML (wie es bei tinyMCE der Fall ist), wird nur `"markup"` geschickt.

### Bereitstellung von `qpy://options/`-URLs in Fragen

Zur Abstraktion der verschiedenen QPy-URL-Typen (es werden ja vermutlich noch mehr werden) habe ich die Klasse `qpy_url_resolver` hinzugefügt, die sowhl URL-Replacement und Serving bündelt und an mehrere `handles_qpy_url_type` delegiert. Aktuell implementieren das Interface `static_file_service` und  `options_file_service`.

## Ersteinmal out-of-scope

- Copy/Paste aus anderen Editoren. (`file_merge_draft_areas`)
- Content-Type einschränken
- Moodle erlaubt auch Unterverzeichnisse im Dateimanager. Das habe ich bisher hardcoded ausgestellt, sollten wir aber irgendwann unterstützen.
- Sollen in WYSIWYG-Editoren statische Dateien zur Verfügung stehen? Aktuell werden da nur `qpy://options`-URLs ersetzt.
- Dateien werden aktuell mit ihrer File-Ref als Dateinamen geserved. Schöner wäre es, den "echten" Dateinamen zu verwenden und diesen (wie im Issue vorgeschlagen) Teil der QPy-URL zu machen.

---

Closes: #201

MHajoha committed:
refactor: move static_file_service into files namespace

MHajoha committed:
refactor(array-converter): specially handle DateTimes

MHajoha committed:
feat: wysiwyg editor with file upload support in options form

MHajoha committed:
feat: resolve and serve options files in question_ui_renderer

MHajoha committed:
feat: add file_upload_element

MHajoha committed:
style: appease moodlecheck

Moodle check doesn't understand array shapes...

MHajoha committed:
test: draft saving in question_service_test

MHajoha committed:
test: add polyfill for PHPUnit withConsecutive

MHajoha committed:
refactor: have render_context take a moodleform again

Specifically requiring a question_edit_form is too much of a headache in UTs.

MHajoha committed:
refactor: move filename inside the file_metadata

This causes fewer problems with PHP auto-casting filenames which just happen to be numerical to ints.

MHajoha committed:
feat: delete options files when question is deleted

MHajoha committed:
feat(array-converter): support unions

MHajoha committed:
feat: implement upload size restrictions

MartinGauk commented on classes/local/api/wysiwyg_editor_data.php:
```diff
@@ -0,0 +1,56 @@
+        public array $files = [],
+        /** @var string|null $html */
+        #[array_key('html')]
+        public ?string $html = null,
```
Du weichst hier vom Issue #199 ab. Für das Paket soll es nur die zwei definierten Felder text und files geben. Weitere LMS-spezifische Felder sind erlaubt, beginnen aber mit Unterstricht (z.B. _textformat).
HTML finde ich nicht so sinnvoll.

MartinGauk commented on classes/local/files/options_file_service.php:
```diff
@@ -0,0 +1,246 @@
+
+    // TODO: When a file is deleted while editing a question, we do not currently delete it from the permanent file area (it just
+    // gets hidden because it's removed from the metadata). We probably should do so in the future, but we must make sure that
+    // previous question versions remain functional _if_ they still exist.
```
Jede Fragenversion speichert die Dateien in einem eigenen Bereich (jede Version hat ihre eigene questionid). Dadurch hat man doch dieses Problem gar nicht!?

MHajoha replied:
Na ist doch super, ich war mir nicht sicher, wie question versions repräsentiert werden, ich hatte nur getestet, dass sie funktionieren und nicht hinterfragt, woran das liegt ^^

MartinGauk commented on classes/local/form/elements/wysiwyg_editor_element.php:
```diff
@@ -0,0 +1,275 @@
+                'subdirs' => self::SUBDIRS,
+                'maxfiles' => $this->fileuploads->maxfiles ?? EDITOR_UNLIMITED_FILES,
+                'maxbytes' => $this->fileuploads->maxbytesperfile ?? FILE_AREA_MAX_BYTES_UNLIMITED,
+                'areamaxbytes' => $this->fileuploads->maxbytestotal ?? FILE_AREA_MAX_BYTES_UNLIMITED,
```
Hier muss `get_max_upload_file_size` verwendet werden.

MartinGauk replied:
Ah, ich sehe gerade, eher `get_user_max_upload_file_size` sollte verwendet werden.

MartinGauk replied:
Wobei ich sehe nun, dass in `\form_filemanager::__construct` schon `get_user_max_upload_file_size` aufgerufen wird. Dann müssen wir uns um das Limit nicht selber kümmern wohl.

MHajoha replied:
Einer der Kommentare in `file_save_draft_area_files` suggeriert, dass es möglich wäre, zumindest `areamaxbytes` zu umgehen...

```php
    // Check if the draft area has exceeded the authorised limit. This should never happen as validation
    // should have taken place before, unless the user is doing something nauthly. If so, let's just not save
    // anything at all in the next area.
```

Da beides dort `maxbytes`, `maxfiles` und `areamaxbytes` nochmal geprüft werden, mache ich das jetzt auch in beiden Elementen im `on_export`: 18bdaef392d593dcfea9c0d2464d24cf90bd6b52

In meinen Tests konnte ich da als freundlicher Nutzer aber auch gar nicht hinkommen, Moodle blockt das vorher mit einer schönen Fehlermeldung. Daher nur `coding_exceptions`.

MHajoha replied:
Tatsächlich sowohl in `form_filemanager` als auch in `MoodleQuickForm_editor` (aber da nur get_max_upload_file_size) / `MoodleQuickForm_filemanager` (aber da ohne `$coursebytes`), ziemlich wirr.

Ich bin etwas unentschlossen (Es würde an sich nicht wehtun, das nochmal zu machen), aber ich würde es auch lassen.

MartinGauk replied:
Ohje, das ist ja merkwürdig. Dann finde ich es schon besser, wenn wir hier und an der anderen Stelle `get_user_max_upload_file_size` benutzen, damit wir sicher sind, dass das Limit dasselbe ist wie bei der Überprüfung in `check_upload_restrictions`.

MHajoha replied:
[9324a1c](https://github.com/questionpy-org/moodle-qtype_questionpy/pull/209/commits/9324a1c745a67aeb58625a46301a59dac5bf82fb)

MartinGauk commented on classes/local/form/elements/file_upload_element.php:
```diff
@@ -0,0 +1,130 @@
+            'maxfiles' => $this->maxfiles ?? EDITOR_UNLIMITED_FILES,
+            // MoodleQuickForm_filemanager doesn't offer a minfiles option, so we leave that to the QPy-side validation.
+            'maxbytes' => $this->maxbytesperfile ?? FILE_AREA_MAX_BYTES_UNLIMITED,
+            'areamaxbytes' => $this->maxbytestotal ?? FILE_AREA_MAX_BYTES_UNLIMITED,
```
Hier muss `get_max_upload_file_size` verwendet werden.

MHajoha replied:
s.o.

MartinGauk commented on classes/local/form/elements/wysiwyg_editor_element.php:
```diff
@@ -0,0 +1,275 @@
+        public bool $includehtml = false,
+        /** @var file_upload_options|disabled_sentinel|null (null means default/no limits, disabled means no file uploads at all) */
+        #[array_key('file_uploads')]
+        public file_upload_options|disabled_sentinel|null $fileuploads = null,
```
Hier finde ich es besser, wenn null disabled bedeutet und kein sentinel verwendet wird.

MartinGauk reviewed (changes requested):
Aufgrund des umfangreichen Änderungen habe ich mir jetzt nicht alles bis ins Detail angesehen und habe es auch nicht ausprobiert.

MartinGauk commented:
> Moodle erlaubt auch Unterverzeichnisse im Dateimanager. Das habe ich bisher hardcoded ausgestellt, sollten wir aber irgendwann unterstützen.

Ich stelle mir mittlerweile als einfachere Lösung vor, dass wir nur eine einfache Liste (keine verschachtelte) an Dateien haben mit jeweils Attribut `path`. Dann sollten wir `path` auch direkt schon in `file_metadata.php` aufnehmen, damit wir keine Daten später migrieren müssen oder an evtl. fehlende Pfade denken müssen.

> Sollen in WYSIWYG-Editoren statische Dateien zur Verfügung stehen? Aktuell werden da nur `qpy://options`-URLs ersetzt.

Ich weiß nicht, ob wir an der Stelle überhaupt QPy-URLs unterstützen sollten.

MartinGauk commented on classes/local/files/options_file_service.php:
```diff
@@ -0,0 +1,246 @@
+     * @throws stored_file_creation_exception
+     * @throws coding_exception
+     */
+    public function save_draft_area_files(int $contextid, int $questionid, int $userid, int $draftitemid): void {
```
Ich bin mir nicht sicher, ob ich noch irgendwas in Moodle übersehen habe, aber ich glaube, dass wir die Dateien noch selber validieren müssen, weil `file_save_draft_area_files` nicht genutzt wird.

MHajoha replied:
Passiert jetzt hier: 18bdaef392d593dcfea9c0d2464d24cf90bd6b52. Wird in den `on_export` aufgerufen, also nicht nur beim speichern, auch, bevor es an den QPy-Server geschickt wird.

In meinen Tests konnte ich da als freundlicher Nutzer nicht hinkommen, Moodle blockt das vorher mit einer schönen Fehlermeldung. Daher nur `coding_exceptions`.

MHajoha committed:
refactor: correct review issues

- Only $text, $textformat and $files in wysiwyg_editor_data
- $path in file_metadata
- $size in file_metadata (not a review issue, but just realized it was missing)
- Change wysiwyg_editor_element->fileuploads semantics: Remove disabled_sentinel, use null for disabled, and new file_upload_options as the default.

MHajoha commented:
> > Moodle erlaubt auch Unterverzeichnisse im Dateimanager. Das habe ich bisher hardcoded ausgestellt, sollten wir aber irgendwann unterstützen.
> 
> Ich stelle mir mittlerweile als einfachere Lösung vor, dass wir nur eine einfache Liste (keine verschachtelte) an Dateien haben mit jeweils Attribut `path`. Dann sollten wir `path` auch direkt schon in `file_metadata.php` aufnehmen, damit wir keine Daten später migrieren müssen oder an evtl. fehlende Pfade denken müssen.

:+1:

Ob und wie wir Informationen über Ordner senden wollen, können wir ja noch schauen, halte ich aber für nicht so wichtig.

> > Sollen in WYSIWYG-Editoren statische Dateien zur Verfügung stehen? Aktuell werden da nur `qpy://options`-URLs ersetzt.
> 
> Ich weiß nicht, ob wir an der Stelle überhaupt QPy-URLs unterstützen sollten.

Aktuell ersetze ich die zum Editor gehörenden `qpy://options/`-URLs, das "müssen" wir auch quasi machen, das ist unser Äquivalent zu Moodle's `@@PLUGINFILE@@`-Ersetzungen. Frage ist, ob man auch URLs auf andere WYSIWYG-Felder oder File-Upload-Felder und/oder auf statische Dateien erlaubt. Ist aber so oder so weit out of scope, denke ich.

MHajoha committed:
feat: check file upload restrictions

MHajoha committed:
chore: remove irrelevant TODO

MartinGauk commented on classes/local/form/elements/wysiwyg_editor_element.php:
```diff
@@ -0,0 +1,272 @@
+        public string $label,
+        /** @var bool */
+        #[array_key('include_html')]
+        public bool $includehtml = false,
```
Hier gibt es noch einen kleinen Überbleibsel.

MHajoha committed:
refactor: delete options files the way god intended

MHajoha committed:
fix: move options files in qtype->move_files

MartinGauk commented:
> Aktuell ersetze ich die zum Editor gehörenden `qpy://options/`-URLs, das "müssen" wir auch quasi machen, das ist unser Äquivalent zu Moodle's `@@PLUGINFILE@@`-Ersetzungen. Frage ist, ob man auch URLs auf andere WYSIWYG-Felder oder File-Upload-Felder und/oder auf statische Dateien erlaubt. Ist aber so oder so weit out of scope, denke ich.

Prinzipiell kann das zwar nützlich sein, solange es sich um Texte von Trainern handelt. Im Falle von Texten von Studis dürfen aber ausschließlich Dateien in dessen Abgabe referenziert und aufgelöst werden. Grundsätzlich will ich, dass Studis auf Options-Files (und im Grunde auch die anderen Dateien außer static) nur zugreifen können, wenn die Datei im Fragentext referenziert wurde.

MHajoha committed:
refactor: call get_user_max_upload_file_size ourselves

Moodle will call it again, but in MoodleQuickForm_filemanager it leaves out $coursebytes, which seems inconsistent. And more calls don't hurt.

MHajoha commented:
> Prinzipiell kann das zwar nützlich sein, solange es sich um Texte von Trainern handelt. Im Falle von Texten von Studis dürfen aber ausschließlich Dateien in dessen Abgabe referenziert und aufgelöst werden. Grundsätzlich will ich, dass Studis auf Options-Files (und im Grunde auch die anderen Dateien außer static) nur zugreifen können, wenn die Datei im Fragentext referenziert wurde.

:+1:, würde ich aber auch erstmal out of scope lassen (also die Einschränkung)?

MartinGauk commented:
Ja, zunächst.
